File: python/Icollections/ListNode.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self):
        ret = []
        while self:
            if self.next is not None:
                ret.append('(' + str(self.val) + '-->' +
                           str(self.next.val) + ')')
            else:
                ret.append('(' + str(self.val) + '--> None )')
            self = self.next
        return ret.__str__()


def convert(arr: list) -> ListNode:
    length = len(arr)
    if not length:
        return None
    nodes = []
    for i in range(length):
        nodes.append(ListNode(arr[i]))
    for i in range(length - 1):
        nodes[i].next = nodes[i + 1]
    return nodes[0]


def reverseList(head: ListNode) -> ListNode:
    def changeNext(pre, head):
        if not head:
            return pre
        next = head.next
        head.next = pre
        pre = head
        return changeNext(pre, next)

    return changeNext(None, head)

File: java/collection/ListNode.java
package collection;

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) {
        val = x;
        next = null;
    }

    // generate a list from an array
    public static ListNode generateList(Object[] array) {
        if (array == null || array.length == 0) {
            return null;
        }
        ListNode head = new ListNode((int) array[0]);
        ListNode node = head;
        for (int i = 1; i < array.length; i++) {
            if (array[i] == null) {
                continue;
            }
            node.next = new ListNode((int) array[i]);
            node = node.next;
        }
        return head;
    }

    // tostring
    public String toString() {
        StringBuilder strb = new StringBuilder();
        ListNode node = this;
        while (node != null) {
            strb.append(node.val + " ");
            node = node.next;
        }
        return strb.toString();
    }

    // reverse list
    public static ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode pre = null;
        ListNode cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
}
